proposals/0000-linear-types.rst

```diff
@@ -0,0 +1,2357 @@
+------------------------
+
+Strict fields, whether unpacked or not, are treated, for the purpose of linearity, just like
+regular fields, *e.g.*
+
+::
+
+    data S a = S !a (S a)
+
+    -- S :: a ->. S a ->. S a
+    --
+    -- Or, polymorphised when used as a term:
+    --
+    -- S :: forall p q. a :p->. S a :q-> S a
+
+::
+
+    data T a = T {-# UNPACK #-}!(a, a) a
+
+    -- T :: (a, a) ->. a ->. T a
+    --
+    -- Or, polymorphised when used as a term:
+    --
+    -- T :: forall p q. (a, a) :p->. a :q-> T a
+
+Base
+----
+
+Because linear functions only strengthen the contract of unrestricted
+functions, a number of functions of ``base`` can get a more precise
+type. However, for pedagogical reasons, to prevent linear types from
+interfering with newcomers' understanding of the ``Prelude``, this
+proposal does not modify ``base``. Instead, we expect that users will
+publish new libraries on Hackage including more precisely typed
+``base`` functions. One such library has already started `here
+<https://github.com/tweag/linear-base>`_.
+
+Any linear variant of ``base`` need not redefine any of the data types
+defined in ``base``. This is because like for all other data types,
+constructors of (non-GADT) data types in ``base`` are linear under
+this proposal. Since we get to reuse data types, libraries
+implementing linear variants of ``base`` functions remain compatible
+with ``base`` (e.g. there need not be two ``Maybe`` types, two list
+types etc).
+
+The only function which will need to change is ``($)`` because its
+typing rule is built in the type checker. Ignoring the details about
+levity and higher-rank polymorphism in the typing rule, the type
+``($)`` will be:
+
+::
+
+  ($) :: (a :p-> b) ⊸ a :p-> b
+
+Defining a linear variant of ``base`` is out of scope of this
+proposal. Possible future standardisation of the library content is
+the competence of the Core Libraries Committee (CLC). For expository
+purposes of the next sections, however, we assume that such a library
+will at least define the following data type:
+
+::
+
+   data Unrestricted a where
+     Unrestricted :: a -> Unrestricted a
+
+See the paper for intuitions about the ``Unrestricted`` data type.
+
+Multiplicities
+--------------
+
+.. _Multiplicities:
+
+So far, we have considered only two multiplicities: ``1`` and ``ω``.
+But the metatheory works with any so-called sup-semi-lattice-ordered
+semi-ring (without a 0) of multiplicities. That is: there is a 1,
+a sum and a product with the usual distributivity laws, a (computable)
+order compatible with the sum and product, such that each pair of
+multiplicities has a (computable) join. Even if there is only two
+multiplicities in this proposal, the proposal is structured to allow
+future extensions.
+
+Here is the definition of sum, product and order for this proposal's
+multiplicities (in Haskell pseudo-syntax):
+
+::
+
+   _ + _ = ω
+
+   1 * x = x
+   x * 1 = x
+   ω * ω = ω
+
+   _ ⩽ ω = True
+   x ⩽ y = x == y
```

Comment:
@yav Sorry to have failed to reply before. Thanks for these points!

>    It should be explicit if each rule strands in its own, or if it assumes the previous ones
>    The second rule for <= seems only correct, assuming that the first does not hold.

What I meant by "Haskell pseudo-syntax" on top was that it read like a regular Haskell definition (in particular the order matters). Maybe that's not clear enough, what would you suggest.

>    I think it is more convenient if each rule stands on its own (i.e, assumptions are explicitly specified)
>    I think it would be better to specify the behavior of <= for each constant: 1 <= x ~ True, x <= inf ~ True, x <= 1 => x ~ 1, inf <= x => x ~ inf
>    With these rules it is obvious that we can always simplify away all constants, and all that's left are products of variables, which should be relatively easy to match

The [Linear Core specification document](https://ghc.haskell.org/trac/ghc/attachment/wiki/LinearTypes/Implementation/minicore.pdf) contains a set of position-independent rules. But it's quite a bit more verbose. It also does not make the assumption that 1 and ω are the only ground multiplicities.

I should say that it is not currently my plan to have the type family unfolding mechanism perform computation over multiplicities, even if they are ground, but rather leave them to the constraint solver.

>    What is the point of having +, if it can always be simplified to inf? (I'll use inf instead of omega as it easier to type)

It is needed by the abstract theory. If the multiplicities are only 1 and ω, then + is, indeed. As mentioned in Section _More Multiplicities_ (current §5.18.2), we'd like to let the door open for other multiplicities, hence the introduction of the abstract theory.

>    Isn't it the case that inf * x= inf for any x? And the other way, it course.

Not if there is a 0 multiplicity (see _Zero as a multiplicity_, currently §5.13), but otherwise yes.




>    I think it is also useful to add x <= x * y, which would allow simplification of common products on each side of a <=

Did you intend to write `x ⩽ y ⟹ x*z ⩽ y * z`? I believe this does hold of any interesting multiplicity semi-ring, and in fact, it is assumed in the [Linear Core specification document](https://ghc.haskell.org/trac/ghc/attachment/wiki/LinearTypes/Implementation/minicore.pdf). Much for the same reason that you state.

Reply:
I think that a general discussion makes sense in a paper, but a proposal should be concrete.  In particular, I thought that this proposal was about adding only two multiplicities: `1` (linear) and `inf` (non-linear).  I think that in that setting you can do more aggressive simplification (e.g., `x <= x * y`, no `+`, etc.)   And it is useful, as the constraint solver can eliminate any constraints containing repeated terms on both sides.

It is quite unclear that we need other multiplicities---you already point other problems with having 0 multiplicities, for example.  I am not sure that we need to pay the price for them now: if in the future we decide that we want them, there could be another proposal, and we could change the rules, or come up with some other solution. 

More generally, I think this proposal would certainly benefit from simplifying away some of the general possibilities, and zooming in on concrete design choices.